From e0c893531995c2217b67ef0059d071ac461f899a Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey.doherty@gmail.com>
Date: Sat, 10 May 2014 18:06:36 +0100
Subject: [PATCH 1/2] Export pkg-config information when building packages

"PkgConfig" fields may be manually added to the Provides of each package
to indicate which pgkconfig files they export, however it is more suitable
to leave this task to PiSi. When building a package that installs a
.pc file, the relevant information is already exported and provided
in the eopkg-index.

This will be followed by changes enabling developers to use these new
pkgconfig exports as build dependencies, greatly reducing package
complexity.
---
 pisi-spec.dtd            |  4 ++++
 pisi-spec.rng            | 13 +++++++++++++
 pisi/metadata.py         |  1 +
 pisi/operations/build.py | 25 +++++++++++++++++++++++++
 pisi/specfile.py         | 12 ++++++++++++
 5 files changed, 55 insertions(+)

diff --git a/pisi-spec.dtd b/pisi-spec.dtd
index 4a84fcb..903edc9 100644
--- a/pisi-spec.dtd
+++ b/pisi-spec.dtd
@@ -108,6 +108,10 @@
 <!ATTLIST COMAR script CDATA #IMPLIED>
 <!ATTLIST COMAR name CDATA #IMPLIED>
 
+<!ELEMENT Provides (PkgConfig)+>
+<!ELEMENT PkgConfig (#PCDATA)>
+<!ATTLIST PkgConfig version CDATA #IMPLIED>
+
 <!ELEMENT Conflicts (Package)+>
 <!-- we have a problem here, this Package tag only contains a package name
 while dtd treats this as another Package section, who said dtd doesnt
diff --git a/pisi-spec.rng b/pisi-spec.rng
index dc17f03..4306ee6 100644
--- a/pisi-spec.rng
+++ b/pisi-spec.rng
@@ -1474,6 +1474,9 @@
             <oneOrMore>
                 <ref name="COMAR"/>
             </oneOrMore>
+            <optional>
+                <ref name="PkgConfig"/>
+            </optional>
         </element>
     </define>
 
@@ -1494,6 +1497,16 @@
         </optional>
     </define>
 
+    <!-- PkgConfig -->
+    <define name="PkgConfig">
+        <element name="PkgConfig">
+            <text/>
+        </element>
+        <optional>
+            <attribute name="version"/>
+        </optional>
+    </define>
+
     <!-- Conflicts -->
     <define name="Conflicts">
         <element name="Conflicts">
diff --git a/pisi/metadata.py b/pisi/metadata.py
index 5c72ca4..341a7c2 100644
--- a/pisi/metadata.py
+++ b/pisi/metadata.py
@@ -123,6 +123,7 @@ class MetaData(xmlfile.XmlFile):
         self.package.conflicts = pkg.conflicts
         self.package.replaces = pkg.replaces
         self.package.providesComar = pkg.providesComar
+        self.package.providesPkgConfig = pkg.providesPkgConfig
         #self.package.requiresComar = pkg.requiresComar
         self.package.additionalFiles = pkg.additionalFiles
 
diff --git a/pisi/operations/build.py b/pisi/operations/build.py
index 1d17df3..53a6b11 100644
--- a/pisi/operations/build.py
+++ b/pisi/operations/build.py
@@ -926,6 +926,31 @@ class Builder:
         for fileinfo in self.files.list:
             size += fileinfo.size
 
+        # Detect any installed pkg-config files
+        pkgconfigExec = pisi.util.search_executable("pkg-config")
+        for fileinfo in self.files.list:
+            path = fileinfo.path
+            if path.endswith(".pc") and "pkgconfig" in path:
+                if not pkgconfigExec:
+                    ctx.ui.warning(_("pkg-config not found but package provides .pc files - skipping export of pkgconfig information"))
+                    break
+                pcName = path.split("/")[-1].split(".pc")[0]
+                alreadyProvided = False
+                # Check its not been manually specified
+                for pkgconfig in metadata.package.providesPkgConfig:
+                    if pkgconfig.om == pcName:
+                        alreadyProvided = True
+                        break
+                if not alreadyProvided:
+                    pkgconfig = pisi.specfile.PkgConfigProvide()
+                    pkgconfig.om = pcName
+                    code,out,err = pisi.util.run_batch("%s --modversion %s" % (pkgconfigExec, pcName))
+                    if code != 0:
+                        ctx.ui.warn(_("Unable to obtain pkgconfig module version for %s") % pcName)
+                    else:
+                        pkgconfig.version = out.strip().rstrip()
+                    metadata.package.providesPkgConfig.append(pkgconfig)
+                    ctx.ui.debug(_("Adding %s to provided pkgconfig list" % pcName))
         metadata.package.installedSize = long(size)
 
         self.metadata = metadata
diff --git a/pisi/specfile.py b/pisi/specfile.py
index b6cdcde..4c6f7a8 100644
--- a/pisi/specfile.py
+++ b/pisi/specfile.py
@@ -150,6 +150,17 @@ class ComarProvide:
         s += ' (' + self.om + '%s' % (' for %s' % self.name if self.name else '') + ')'
         return s
 
+class PkgConfigProvide:
+
+    s_om = [autoxml.String, autoxml.mandatory]
+    a_version = [autoxml.String, autoxml.optional]
+
+    def __str__(self):
+        s = self.om
+        if self.a_version and self.a_version != '':
+            s += " == " + self.a_version
+        return s
+
 class Archive:
 
     s_uri = [ autoxml.String, autoxml.mandatory ]
@@ -243,6 +254,7 @@ class Package:
     t_Conflicts = [ [pisi.conflict.Conflict], autoxml.optional, "Conflicts/Package"]
     t_Replaces = [ [pisi.replace.Replace], autoxml.optional, "Replaces/Package"]
     t_ProvidesComar = [ [ComarProvide], autoxml.optional, "Provides/COMAR"]
+    t_ProvidesPkgConfig = [ [PkgConfigProvide], autoxml.optional, "Provides/PkgConfig"]
     t_AdditionalFiles = [ [AdditionalFile], autoxml.optional]
     t_History = [ [Update], autoxml.optional]
 
-- 
1.9.1

